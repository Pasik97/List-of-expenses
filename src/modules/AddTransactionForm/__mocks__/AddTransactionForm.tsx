const mock = () => '(AddTransactionForm /)';

export default mock;
