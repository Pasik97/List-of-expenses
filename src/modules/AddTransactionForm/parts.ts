import styled from 'styled-components';
import media from 'utils/media';

export const AddTransactionFormWrapper = styled.div`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   margin: 24px 0;
`;

export const Row = styled.div`
   margin-bottom: 8px;

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 3fr 1fr;
   }
`;

export const InputWrapper = styled.div`
   margin: 0 0 8px 0;

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 1fr 1fr;
   }
`;

export const InputLabel = styled.p`
   white-space: nowrap;
   margin: 0;

   ${media.onlyMobile} {
      margin-bottom: 6px;
   }

   ${media.fromTablet} {
      flex-grow: 1;
   }
`;

export const Input = styled.input`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   border: none;
   outline: none;
   border-bottom: 1px solid #dddddd;

   &:focus {
      border-bottom: 1px solid #1a1a1a;
   }

   ${media.fromTablet} {
      flex-grow: 1;
   }
`;

export const AddButton = styled.button`
   color: #1a1a1a;
   font-size: 18px;
   line-height: 21px;
   background: none;
   border: 2px solid #dddddd;
   padding: 6px 24px;

   &:disabled {
      color: #dddddd;
   }

   &:hover, &active {
      border-color: #1a1a1a;
      cursor: pointer;
   }

   ${media.fromTablet} {
      height: fit-content;
      width: fit-content;
      justify-self: flex-end;
   }
`;

export const ValidateInputWrapper = styled.div`
   display: flex;
   flex-direction: column;

   ${media.fromTablet} {
      margin-left: 16px;
      flex-grow: 1;
   }
`;

export const Validation = styled.p`
   margin: 8px 0 0 0;
   color: red;
   font-size: 13px;
   line-height: 16px;
   min-height: 16px;
`;
