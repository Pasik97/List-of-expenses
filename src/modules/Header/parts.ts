import styled from 'styled-components';

export const HeaderWrapper = styled.header`
   display: flex;
   align-items: baseline;
   justify-content: space-between;
   margin: 16px 0;
`;

export const Title = styled.h2`
   color: #1a1a1a;
   font-size: 28px;
   line-height: 32px;
   margin: 0 8px 0 0;
   font-weight: 700;
`;

export const ConversionRateWrapper = styled.div`
   display: flex;
   align-items: baseline;
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
`;

export const RateInput = styled.input`
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
   padding: 0;
   border: none;
   outline: none;
   border-bottom: 1px solid #dddddd;
   width: 60px;

   &:focus {
      border-bottom: 1px solid #1a1a1a;
   }
`;

export const ConversionLabel = styled.p`
   margin: 0;
   white-space: pre;
`;
