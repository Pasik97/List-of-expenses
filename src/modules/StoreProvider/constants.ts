import { ReactNode } from "react";
import Stores from "stores/Stores";

export interface StoreProviderProps {
   store: Stores;
   children: ReactNode;
}
