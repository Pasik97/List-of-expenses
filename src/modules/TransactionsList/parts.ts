import styled from 'styled-components';
import media from 'utils/media';

export const TransactionsListWrapper = styled.div``;

export const Cell = styled.div`
   color: #1a1a1a;
   font-size: 16px;
   line-height: 21px;
   border-right: 1px solid #1a1a1a;
   padding: 6px;
   display: flex;
   align-items: center;

   &:last-of-type {
      border-right: none;
   }
`;

export const HeaderCell = styled(Cell)`
   font-weight: 700;
   background-color: #cccccc;
   
   &:hover, &:active {
      cursor: pointer;
   }
`;

export const Row = styled.div`
   display: grid;
   grid-template-columns: 1fr 1fr 1fr 1fr;
   border: 1px solid #1a1a1a;
   border-top: none;

   &:first-of-type {
      border-top: 1px solid #1a1a1a;
   }

   &:nth-of-type(2n+3) {
      ${Cell} {
         background-color: #eeeeee;
      }
   }

   ${media.fromTablet} {
      display: grid;
      grid-template-columns: 2fr 1fr 1fr 1fr;
   }

   ${media.fromDesktop} {
      display: grid;
      grid-template-columns: 3fr 1fr 1fr 1fr;
   }
`;

export const Label = styled.p`
   margin: 0;
   overflow-wrap: anywhere;
`;

export const DeleteButton = styled.button`
   color: #4d4d4d;
   font-size: 16px;
   line-height: 21px;
   border: none;
   outline: none;
   background: none;

   &:hover, &:active {
      cursor: pointer;
      color: #1a1a1a;
   }
`;
