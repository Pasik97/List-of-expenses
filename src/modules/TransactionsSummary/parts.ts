import styled from 'styled-components';

export const Summary = styled.p`
   color: #1a1a1a;
   font-size: 18px; 
   line-height: 21px;
   margin: 16px 0 0 0;
`;
