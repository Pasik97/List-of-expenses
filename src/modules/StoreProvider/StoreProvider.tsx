import React, { createContext, } from 'react';
import { StoreProviderProps } from './constants';
import Stores from 'stores/Stores';

export const StoreProviderContext = createContext<Stores>({} as Stores);

const StoreProvider: React.FC<StoreProviderProps> = ({
   children,
   store,
}) => (
   <StoreProviderContext.Provider value={store}>
      {children}
   </StoreProviderContext.Provider>
);

export default StoreProvider;
