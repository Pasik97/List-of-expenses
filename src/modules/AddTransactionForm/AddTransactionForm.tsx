import React, { useContext, useState } from 'react';
import { observer } from 'mobx-react';
import * as P from './parts';
import { StoreProviderContext } from 'modules/StoreProvider/StoreProvider';
import { doubleNumberRegex } from 'utils/validators';
import { InitialTransactionData } from 'stores/transactions/constants';

const AddTransactionForm: React.FC = observer(() => {
   const { transactions } = useContext(StoreProviderContext);

   const [title, setTitle] = useState<string>('');
   const [titleValidation, setTitleValidation] = useState<string>('');
   const [amount, setAmount] = useState<string>('');
   const [amountValidation, setAmountValidation] = useState<string>('');

   const onTitleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      setTitle(event.target.value);

      const textToValidate = event.target.value.trim();

      if (!textToValidate.length) {
         setTitleValidation('Field required');
      } else if (textToValidate.length < 5) {
         setTitleValidation('Title should have at least 5 characters');
      } else {
         setTitleValidation('');
      }
   };

   const onAmountChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const isNumberValid = doubleNumberRegex.test(event.target.value);
      const decimals = event.target.value.split('.')[1];
      const areDecimalsValid = !decimals || decimals.length <= 2;

      if (!event.target.value.length || (isNumberValid && areDecimalsValid)) {
         setAmount(event.target.value);

         if (!event.target.value.length) {
            setAmountValidation('Field required');
         } else {
            setAmountValidation('');
         }
      }
   };

   const onAddButtonClick = () => {
      const newTransaction: InitialTransactionData = {
         name: title,
         amountPln: Number(amount),
      };

      transactions.addNewTransaction(newTransaction);
      setTitle('');
      setAmount('');
   };

   const isButtonDisabled = !transactions.conversionRate 
   || !!titleValidation 
   || !!amountValidation
   || !title
   || !amount;

   return (
      <P.AddTransactionFormWrapper>
         <P.Row>
            <P.InputWrapper>
               <P.InputLabel>Title of transaction</P.InputLabel>
               <P.ValidateInputWrapper>
                  <P.Input name="title" value={title} onChange={onTitleChange} />
                  <P.Validation>{titleValidation}</P.Validation>
               </P.ValidateInputWrapper>
            </P.InputWrapper>
         </P.Row>
         <P.Row>
            <P.InputWrapper>
               <P.InputLabel>Amount (in PLN) </P.InputLabel>
               <P.ValidateInputWrapper>
                  <P.Input name="amount" value={amount} onChange={onAmountChange} />
                  <P.Validation>{amountValidation}</P.Validation>
               </P.ValidateInputWrapper>
            </P.InputWrapper>
            <P.AddButton disabled={isButtonDisabled} onClick={onAddButtonClick}>Add</P.AddButton>
         </P.Row>
      </P.AddTransactionFormWrapper>
   )
});


export default AddTransactionForm;