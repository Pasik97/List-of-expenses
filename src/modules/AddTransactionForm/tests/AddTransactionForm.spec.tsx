import { act, create } from "react-test-renderer";
import * as M from './mockData';
import AddTransactionForm from "../AddTransactionForm";
import { AddButton, Input, Validation } from "../parts";
import StoreProvider from "modules/StoreProvider/StoreProvider";

describe('<AddTransactionForm />', () => {
   it('should render button and 2 inputs to enter transaction title and amount', () => {
      const component = create(
         <StoreProvider store={M.mockStore}>
            <AddTransactionForm />
         </StoreProvider>,
      );

      expect(component.toJSON()).toMatchSnapshot();
   });

   describe('onTitleChange', () => {
      it('should set title input value if user enter title longer than 5 characters', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let titleInput = component.root.findAllByType(Input)[0];

         act(() => {
            titleInput.props.onChange({ target: { value: 'Mock title' } });
         });

         titleInput = component.root.findAllByType(Input)[0];
   
         expect(titleInput.props.value).toEqual('Mock title');
      });

      it('should set title input value and set title validation if user enter title shorter than 5 characters', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let titleInput = component.root.findAllByType(Input)[0];
         let validationTitle = component.root.findAllByType(Validation)[0];
         
         expect(validationTitle.props.children).toEqual('');

         act(() => {
            titleInput.props.onChange({ target: { value: '  a  ' } });
         });

         titleInput = component.root.findAllByType(Input)[0];
         validationTitle = component.root.findAllByType(Validation)[0];
   
         expect(titleInput.props.value).toEqual('  a  ');
         expect(validationTitle.props.children).toEqual('Title should have at least 5 characters');
      });

      it('should set title input value and set title validation if user clear input', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let titleInput = component.root.findAllByType(Input)[0];
         let validationTitle = component.root.findAllByType(Validation)[0];
         
         expect(validationTitle.props.children).toEqual('');

         act(() => {
            titleInput.props.onChange({ target: { value: '' } });
         });

         titleInput = component.root.findAllByType(Input)[0];
         validationTitle = component.root.findAllByType(Validation)[0];
   
         expect(titleInput.props.value).toEqual('');
         expect(validationTitle.props.children).toEqual('Field required');
      });
   });

   describe('onAmountChange', () => {
      it('should set amount input value if user enter valid number', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let amountInput = component.root.findAllByType(Input)[1];

         act(() => {
            amountInput.props.onChange({ target: { value: '12.11' } });
         });

         amountInput = component.root.findAllByType(Input)[1];
   
         expect(amountInput.props.value).toEqual('12.11');
      });

      it('should set amount input value and amount validation if user clear input', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let amountInput = component.root.findAllByType(Input)[1];
         let validationAmount = component.root.findAllByType(Validation)[1];
         
         expect(validationAmount.props.children).toEqual('');

         act(() => {
            amountInput.props.onChange({ target: { value: '' } });
         });

         amountInput = component.root.findAllByType(Input)[1];
         validationAmount = component.root.findAllByType(Validation)[1];
   
         expect(amountInput.props.value).toEqual('');
         expect(validationAmount.props.children).toEqual('Field required');
      });

      it('should not set amount input value if user enter invalid value', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let amountInput = component.root.findAllByType(Input)[1];

         act(() => {
            amountInput.props.onChange({ target: { value: 'awdawd' } });
         });

         amountInput = component.root.findAllByType(Input)[1];
   
         expect(amountInput.props.value).toEqual('');
      });

      it('should not set amount input value if decimals in number entered by user are longer than 2', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let amountInput = component.root.findAllByType(Input)[1];

         act(() => {
            amountInput.props.onChange({ target: { value: '12.2222222' } });
         });

         amountInput = component.root.findAllByType(Input)[1];
   
         expect(amountInput.props.value).toEqual('');
      });
   });

   describe('onAddButtonClick', () => {
      it('should call addNewTransaction Transactions class method', () => {
         const component = create(
            <StoreProvider store={M.mockStore}>
               <AddTransactionForm />
            </StoreProvider>,
         );
         let addButton = component.root.findByType(AddButton);

         addButton.props.onClick();
   
         expect(M.mockStore.transactions.addNewTransaction).toHaveBeenCalled();
      });
   });
});
