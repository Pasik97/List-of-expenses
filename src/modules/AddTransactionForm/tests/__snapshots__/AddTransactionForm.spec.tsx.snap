// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddTransactionForm /> should render button and 2 inputs to enter transaction title and amount 1`] = `
<div
  className="sc-bdvvtL hwnKLg"
>
  <div
    className="sc-gsDKAQ ALJyz"
  >
    <div
      className="sc-dkPtRN kIKkZP"
    >
      <p
        className="sc-hKwDye gaRtAI"
      >
        Title of transaction
      </p>
      <div
        className="sc-gKclnd ccezja"
      >
        <input
          className="sc-eCImPb ZGNVb"
          name="title"
          onChange={[Function]}
          value=""
        />
        <p
          className="sc-iCfMLu ftLxkj"
        >
          
        </p>
      </div>
    </div>
  </div>
  <div
    className="sc-gsDKAQ ALJyz"
  >
    <div
      className="sc-dkPtRN kIKkZP"
    >
      <p
        className="sc-hKwDye gaRtAI"
      >
        Amount (in PLN) 
      </p>
      <div
        className="sc-gKclnd ccezja"
      >
        <input
          className="sc-eCImPb ZGNVb"
          name="amount"
          onChange={[Function]}
          value=""
        />
        <p
          className="sc-iCfMLu ftLxkj"
        >
          
        </p>
      </div>
    </div>
    <button
      className="sc-jRQBWg cDXwSo"
      disabled={true}
      onClick={[Function]}
    >
      Add
    </button>
  </div>
</div>
`;
